### PR TITLE
Automatic transfer of active fe indices.

### PR DESCRIPTION
**Pull request by marcfehling**
This pull request introduces the automatic transfer of each cell's `active_fe_index` during refinement and coarsening whenever an `hp::DoFHandler` is used with any kind of `Triangulation`. This is done by connecting dedicated functions to the corresponding signals of the underlying `Triangulation` object.

Class `p::d::ActiveFEIndicesTransfer` introduced in #6961 is now obsolete. Hence it will be removed in this pull request, since it was never part of an official `deal.II` release.

I am not completely happy with serialization concerning`hp::DoFHandler` and `p::d::Triangulation`, since the current implementation with `handles` requires everything to happen in order. However, we will always call `Triangulation::save()` last and `Triangulation::load()` first in the sequence of registering and unpacking. Thus, I introduced two functions `hp::DoFHandler::prepare_for_serialization_of_active_fe_indices()` and `hp::DoFHandler::deserialize_active_fe_indices()` with which serialization and deserialization can be triggered manually. We may think about this in the future.

Part of #3511.

### CODE REVIEW

File: include/deal.II/hp/dof_handler.h [hunk 1]
```diff
@@ -884,6 +898,39 @@ namespace hp
...
+     * the active_fe_indices on all active cells as well.
+     *
+     * This function registers that these indices are to be stored whenever the
+     * parallel::distributed::Triangulation::save() function is called on the
+     * underlying triangulation.
+     *
+     * @note Currently only implemented for triangulations of type
+     *   parallel::distributed::Triangulation.
+     *
+     * @note See parallel::distributed::SolutionTransfer for further information.
```
Comment (bangerth) [inline include/deal.II/hp/dof_handler.h]: I think you can use `@see` here. You may want to check how that is rendered to know what sentence structure is useful.

(In reality, I think both ways are fine and it's ok if you don't want to make the change. I just wanted to point out the existence of `@see` -- http://www.doxygen.nl/manual/commands.html#cmdsee and the compiled example here: http://www.doxygen.nl/manual/examples/autolink/html/class_autolink___test.html)

File: include/deal.II/hp/dof_handler.h [hunk 2]
```diff
@@ -884,6 +898,39 @@ namespace hp
...
+    void
+    prepare_for_serialization_of_active_fe_indices();
+
+    /**
+     * Whenever serialization with a parallel::distributed::Triangulation as the
+     * underlying triangulation is considered, we also need to consider storing
+     * the active_fe_indices on all active cells as well.
+     *
+     * This function deserializes and distributes the previously stored
+     * active_fe_indices on all active cells.
```
Comment (bangerth) [inline include/deal.II/hp/dof_handler.h]: same comments here

File: include/deal.II/hp/dof_handler.h [hunk 3]
```diff
@@ -1120,6 +1220,22 @@ namespace hp
      */
     std::map<const cell_iterator, const unsigned int> coarsened_cells_fe_index;
 
+    /**
+     * Container to temporarily store the active_fe_index of every locally
+     * owned cell for transfer across distributed triangulations.
+     */
+    std::vector<unsigned int> active_fe_indices;
```
Comment (bangerth) [inline include/deal.II/hp/dof_handler.h]: Can you open a github issue that reminds us to put all of these temporary objects into some kind of structure? They all have perfectly innocuous names that suggest that they store some kind of important information, but in reality they are all only present and valid during very specific phases of the program. We should mark this somehow by providing them a separate home.

Not for this PR, but if you open a separate issue we can do this at a later time -- feel free to copy the comment above.

File: source/hp/dof_handler.cc [hunk 4]
```diff
@@ -1349,7 +1349,8 @@ namespace hp
     // use the cell-wise access.
     active_cell_iterator cell = begin_active(), endc = end();
     for (unsigned int i = 0; cell != endc; ++cell, ++i)
-      active_fe_indices[i] = cell->active_fe_index();
+      if (cell->is_locally_owned())
+        active_fe_indices[i] = cell->active_fe_index();
```
Comment (bangerth) [inline source/hp/dof_handler.cc]: @masterleinad has tried very hard to convert all of these loops into range-based for loops. Can you do the same here?

File: source/hp/dof_handler.cc [hunk 5]
```diff
@@ -1522,18 +1505,71 @@ namespace hp
...
-    // p::d::Triangulation. In this case, the class ActiveFEIndicesTransfer
-    // has to be consulted.
-    if (dynamic_cast<const parallel::distributed::Triangulation<dim, spacedim>
-                       *>(&*this->tria) == nullptr)
+    // decide whether we need a sequential or a parallel shared/distributed
+    // policy and attach corresponding callback functions dealing with the
+    // transfer of active_fe_indices
+    if (const auto *distributed_tria = dynamic_cast<
+          const parallel::distributed::Triangulation<dim, spacedim> *>(
+          &*this->tria))
```
Comment (bangerth) [inline source/hp/dof_handler.cc]: awkward -- can't you use `&this->get_triangulation()` instead of `&*tria`?

File: source/hp/dof_handler.cc [hunk 6]
```diff
@@ -1868,19 +1972,113 @@ namespace hp
...
           }
 
+        // Clear stored active_fe_indices.
+        refined_cells_fe_index.clear();
+        coarsened_cells_fe_index.clear();
+      }
+  }
+
+
+  template <int dim, int spacedim>
```
Comment (bangerth) [inline source/hp/dof_handler.cc]: three empty lines between functions

File: source/hp/dof_handler.cc [hunk 7]
```diff
@@ -1868,19 +1972,113 @@ namespace hp
...
+  void
+  DoFHandler<dim, spacedim>::post_shared_refinement_fe_index_transfer()
+  {
+    // Finite elements need to be assigned to each cell by calling
+    // distribute_dofs() first to make this functionality available.
+    if (fe_collection.size() > 0)
```
Comment (bangerth) [inline source/hp/dof_handler.cc]: oh, I see the answer to my questions earlier on. This is awkward. You might want to explain in the code snippet I annotated why it is necessary to distribute dof indices...

Comment (bangerth) [reply source/hp/dof_handler.cc]: I guess what we really need is a `set_fe()` function that just sets the FE collection but doesn't actually distribute DoFs -- but that's a substantially larger change to ensure that things continue to be internally consistent. So maybe don't do this right now...

File: include/deal.II/hp/dof_handler.h [hunk 8]
```diff
@@ -884,6 +898,39 @@ namespace hp
     virtual std::size_t
     memory_consumption() const;
 
+    /**
+     * Whenever serialization with a parallel::distributed::Triangulation as the
+     * underlying triangulation is considered, we also need to consider storing
+     * the active_fe_indices on all active cells as well.
```
[the remark this answers is not in the document]
Comment (marcfehling) [reply include/deal.II/hp/dof_handler.h]: I did not find any corresponding entry in the [glossary](https://www.dealii.org/9.0.0/doxygen/deal.II/DEALGlossary.html).

I could provide one (or at least a draft), maybe in a separate pull request?

File: include/deal.II/hp/dof_handler.h [hunk 9]
```diff
@@ -884,6 +898,39 @@ namespace hp
...
+     *
+     * This function registers that these indices are to be stored whenever the
+     * parallel::distributed::Triangulation::save() function is called on the
+     * underlying triangulation.
+     *
+     * @note Currently only implemented for triangulations of type
+     *   parallel::distributed::Triangulation.
```
[the remark this answers is not in the document]
Comment (marcfehling) [reply include/deal.II/hp/dof_handler.h]: An assertion will be triggered.

I'll clarify this with an additional comment.

File: include/deal.II/hp/dof_handler.h (hunk 1 above)
Comment (marcfehling) [reply include/deal.II/hp/dof_handler.h]: I'll give it a try and post a screenshot here later.

File: source/hp/dof_handler.cc (hunk 7 above)
Comment (marcfehling) [reply source/hp/dof_handler.cc]: A function like that would be handy for this case. We need to attach the `FECollection` before assigning `active_fe_indices` in any case.

I'll open an issue for this.

File: include/deal.II/hp/dof_handler.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/hp/dof_handler.h]: Nah, don't worry about it then.

File: include/deal.II/hp/dof_handler.h (hunk 1 above)
Comment (marcfehling) [reply include/deal.II/hp/dof_handler.h]: You agree with this?
doxygen-hp-serialization

Comment (bangerth) [reply include/deal.II/hp/dof_handler.h]: It looks a bit clunky with the grammar. How about write it as this:
```
  @see The documentation of X::Y has further information on serialization.
```

**Comment by marcfehling** [comment]
I just realized that we also need to transfer the `active_fe_indices` when calling `p::d::Triangulation::repartition`.

Shall we trigger the `pre_distributed_refinement` and `post_distributed_refinement` signals in this case as well?

**Comment by marcfehling** [comment]
There are two tests `hp_cell_weights_01` and `hp_cell_weights_02` that rely on `p::d::Triangulation::repartition()`. However, those tests do not fail, presumably because the ownership of the one cell does not change.

I'll modify those tests once we figured out how to deal with the `repartition()` function.

**Comment by marcfehling** [comment]
Further, I didn't pay attention to compilation without `MPI` and `P4EST`. I added corresponding checks, but the code looks rather bulky now. We could move these `pre/post_refinement` functions into the corresponding policy, if you agree.

We would have to extend the `PolicyBase` class with the corresponding `pre/post_refinement` functions. And the `ParallelDistributed` class would get a `unique_ptr` to a `p::d::CellDataTransfer` object as a class member, which will only be initialized when a `hp::DoFHandler` is provided.

What do you think about that?

File: source/hp/dof_handler.cc [hunk 10]
```diff
@@ -1522,22 +1520,78 @@ namespace hp
...
-                    std::ref(*this))));
-        tria_listeners.push_back(this->tria->signals.post_refinement.connect(
-          std::bind(&DoFHandler<dim, spacedim>::post_refinement_fe_index_update,
-                    std::ref(*this))));
+        policy = std_cxx14::make_unique<
+          internal::DoFHandlerImplementation::Policy::ParallelDistributed<
+            DoFHandler<dim, spacedim>>>(*this);
+
+#ifdef DEAL_II_WITH_P4EST
+        cell_data_transfer = std_cxx14::make_unique<
```
Comment (bangerth) [inline source/hp/dof_handler.cc]: why is this necessary? does it not compile if you don't have P4EST configured?

File: source/hp/dof_handler.cc [hunk 11]
```diff
@@ -1747,7 +1872,7 @@ namespace hp
             &(*tria)));
 
         std::vector<types::subdomain_id> saved_subdomain_ids;
-        if (shared_tria != nullptr && shared_tria->with_artificial_cells())
+        if (shared_tria->with_artificial_cells())
```
Comment (bangerth) [inline source/hp/dof_handler.cc]: You might want to put an `Assert(shared_tria != nullptr, ExcInternalError());` right after the `dynamic_cast` (which we know should succeed, but better check).

File: include/deal.II/hp/dof_handler.h [hunk 12]
```diff
@@ -1120,6 +1224,25 @@ namespace hp
      */
     std::map<const cell_iterator, const unsigned int> coarsened_cells_fe_index;
 
+#ifdef DEAL_II_WITH_P4EST
+    /**
```
Comment (bangerth) [inline include/deal.II/hp/dof_handler.h]: Let's not do this `#ifdef` here. If deal.II wasn't configured with p4est, then these member variables will simply not ever be used. But at least the size of the object is always the same.

Comment (marcfehling) [reply include/deal.II/hp/dof_handler.h]: I haven't looked at it from this perspective. I agree.

File: source/hp/dof_handler.cc (hunk 10 above)
Comment (marcfehling) [reply source/hp/dof_handler.cc]: The implementation of the `CellDataTransfer` class is currently embedded in a corresponding `DEAL_II_WITH_P4EST` preprocessor conditional, including its constructor.

The linker complains when `deal.II` is not configured with `p4est`.

**Comment by marcfehling** [comment]
Applied suggestions, squashed all commits into one. Ready for approval and merge once tests succeed.

Some additional (smaller) tasks emerged from the initial PR, postponed in several issues. I need to see when I find time to work them off.
